File: backend/src/main/java/hu/bsstudio/raktr/model/Ticket.java
package hu.bsstudio.raktr.model;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import hu.bsstudio.raktr.dal.entity.Comment;
import hu.bsstudio.raktr.dal.entity.User;
import hu.bsstudio.raktr.dal.value.ProblemSeverity;
import hu.bsstudio.raktr.dal.value.TicketStatus;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import java.util.Date;
import java.util.List;

@EqualsAndHashCode()
@Entity
@Table(name = "ticket")
@JsonDeserialize(builder = Ticket.TicketBuilder.class)
@NoArgsConstructor
@AllArgsConstructor
@ToString
@Data
public class Ticket {

    @Id
    @Setter(AccessLevel.NONE)
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    private String body;

    @NotNull
    private Date dateOfWriting;

    @ManyToOne(fetch = FetchType.LAZY)
    @NotNull
    private User writer;

    @NotNull
    private TicketStatus status;

    @ManyToOne
    private Scannable scannableOfProblem;

    @NotNull
    private ProblemSeverity severity;

    @OneToMany(fetch = FetchType.LAZY)
    private List<Comment> comments;

    public void addComment(final Comment comment) {
        comments.add(comment);
    }

    public void removeComment(final Comment comment) {
        comments.remove(comment);
    }

    @JsonPOJOBuilder(withPrefix = "")
    public static class TicketBuilder {
    }
}

File: frontend/src/app/_model/Ticket.ts
import {Commentable} from './Commentable';
import {ProblemSeverity} from './ProblemSeverity';
import {Scannable} from './Scannable';
import {User} from './User';
import {TicketStatus} from './TicketStatus';
import {Device} from './Device';
import {CompositeItem} from './CompositeItem';
import {Comment} from './Comment';

export class Ticket extends Commentable {
    status: TicketStatus;
    scannableOfProblem: Scannable;
    severity: ProblemSeverity;
    comments: Comment[];

    static fromJson(ticket: Ticket): Ticket {
        const newTicket = new Ticket(
            ticket.id,
            ticket.body,
            new Date(ticket.dateOfWriting),
            ticket.writer,
            ticket.status,
            null,
            ticket.severity
        );

        if (ticket.scannableOfProblem !== undefined) {
            if (ticket.scannableOfProblem['@type'] === 'device') {
                newTicket.scannableOfProblem = Device.fromJson(ticket.scannableOfProblem as Device);
            } else if (ticket.scannableOfProblem['@type'] === 'compositeItem') {
                newTicket.scannableOfProblem = CompositeItem.fromJson(ticket.scannableOfProblem as CompositeItem);
            }
        }

        ticket.comments.forEach(rawComment => {
            newTicket.comments.push(Comment.fromJson(rawComment));
        })

        return newTicket;
    }

    constructor(id: number = -1, body: string = '', dateOfWriting: Date = new Date(), writer: User = null, status: TicketStatus = TicketStatus.OPEN,
                scannableOfProblem: Scannable = null, severity: ProblemSeverity = ProblemSeverity.SEVERE, comments: Comment[] = []) {
        super('ticket', id, body, dateOfWriting, writer);
        this.status = status;
        this.scannableOfProblem = scannableOfProblem;
        this.severity = severity;
        this.comments = comments;
    }

    toJson(): string {
        return `{\"Ticket\": ${this.toJsonWithoutRoot()}}`;
    }

    toJsonWithoutRoot(): string {
        const ticketJson = JSON.parse(JSON.stringify(this));
        ticketJson['@type'] = 'ticket';
        ticketJson['scannableOfProblem'] = JSON.parse(this.scannableOfProblem.toJsonWithoutRoot());

        if (ticketJson.comments !== undefined && ticketJson.comments.length !== 0) {
            ticketJson.comments.forEach(comment => {
                comment['@type'] = comment['type_'];
            })
        }

        return JSON.stringify(ticketJson);
    }
}
